[DTensor] Log API usage metrics for DTensor and DeviceMesh

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167375
* __->__ #167374

Logging propagate_op_sharding_non_cached is a compromise between
 - logging in DTensor.__init__ to catch ALL DTensor usage
 - sparing the overhead in a latency-senstitive region like
   DTensor.__init__
 - and 'real' DTensor usage should incur at least one call to sharding
   propagation

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta @msaroufim @dcci

wconstab committed:
[DTensor] Log API usage metrics for DTensor and DeviceMesh

Logging propagate_op_sharding_non_cached is a compromise between
 - logging in DTensor.__init__ to catch ALL DTensor usage
 - sparing the overhead in a latency-senstitive region like
   DTensor.__init__
 - and 'real' DTensor usage should incur at least one call to sharding
   propagation

[ghstack-poisoned]

zpcore reviewed (approved):
LGTM!

zpcore commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -348,6 +348,8 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
         Propagate the sharding for an operator given the op_schema.
         """
+        # no-op in OSS, logs API usage metrics in meta-internal runs
+        torch._C._log_api_usage_once("torch.distributed._tensor.propagate_op_sharding_non_cached")
```
Maybe add a small test to see if it works or not?

wconstab committed:
Update on "[DTensor] Log API usage metrics for DTensor and DeviceMesh"


Logging propagate_op_sharding_non_cached is a compromise between
 - logging in DTensor.__init__ to catch ALL DTensor usage
 - sparing the overhead in a latency-senstitive region like
   DTensor.__init__
 - and 'real' DTensor usage should incur at least one call to sharding
   propagation

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
